### PR TITLE
Rename should be a no-op if the name is unchanged

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/36063

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs [hunk 1]
```diff
@@ -674,14 +680,21 @@ private void Cancel(bool rollbackTemporaryEdits)
         }
 
         public void Commit(bool previewChanges = false)
+            => CommitWorker(previewChanges);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs]: What's with the worker method here?

File: src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs [hunk 2]
```diff
@@ -843,5 +860,19 @@ internal bool TryGetContainingEditableSpan(SnapshotPoint point, out SnapshotSpan
...
+        public struct TestAccessor
+        {
+            private readonly InlineRenameSession _inlineRenameSession;
+
+            public TestAccessor(InlineRenameSession inlineRenameSession)
+                => _inlineRenameSession = inlineRenameSession;
+
+            public bool CommitWorker(bool previewChanges)
+                => _inlineRenameSession.CommitWorker(previewChanges);
+        }
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs]: What is this all buying us here? It seems like we went from a nice public Commit method to something that's internal and now we have a TestAccessor for it?

**Comment by jasonmalinowski** [comment]
Also @CyrusNajmabadi is this better done as a fast deeper in the engine instead? If I start typing say new letters at the end of the name and then go back to the original name that shouldn't be running the full conflict resolution either. It seems like if we add a fast path to the core rename engine to say the Solution snapshot in and out in that case is identical, then this should also be a fast no-op since the commit will just not apply anything new.

File: src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs]: i would like to validate from tests taht we actually canceled.

**Comment by CyrusNajmabadi** [comment]
>  It seems like if we add a fast path to the core rename engine to say the Solution snapshot in and out in that case is identical, then this should also be a fast no-op since the commit will just not apply anything new.

I can try that.  Thanks!

**Comment by CyrusNajmabadi** [comment]
Talked offline with @jasonmalinowski .  Going to stick with the simpler approach.  Have commented why we think it's reasonable.

File: src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs [hunk 3]
```diff
@@ -674,14 +680,29 @@ private void Cancel(bool rollbackTemporaryEdits)
...
             VerifyNotDismissed();
 
-            if (this.ReplacementText == string.Empty)
+            // If the identifier was deleted (or didn't change at all) then cancel the operation.
+            // Note: an alternative approach would be for the work we're doing (like detecting
+            // conflicts) to quickly bail in the case of no change.  However, that involves deeper
+            // changes to the system and is less easy to validate that nothing happens.
+            //
+            // The only potential downside here would be if there was a language that wanted to
+            // still 'rename' even if the identifier went away (or was unchanged).  But that  isn't
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs]:
```suggestion
            // still 'rename' even if the identifier went away (or was unchanged).  But that isn't
```

File: src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs [hunk 4]
```diff
@@ -674,14 +680,29 @@ private void Cancel(bool rollbackTemporaryEdits)
...
+        private bool CommitWorker(bool previewChanges)
         {
             AssertIsForeground();
             VerifyNotDismissed();
 
-            if (this.ReplacementText == string.Empty)
+            // If the identifier was deleted (or didn't change at all) then cancel the operation.
+            // Note: an alternative approach would be for the work we're doing (like detecting
+            // conflicts) to quickly bail in the case of no change.  However, that involves deeper
+            // changes to the system and is less easy to validate that nothing happens.
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs]: Might be worth calling out the benefit for that approach would be less processing to update the dashboard, but that's the only real difference.

File: src/EditorFeatures/Test2/Rename/InlineRenameTests.vb [hunk 5]
```diff
@@ -1868,5 +1868,39 @@ class [|$$Test1|]
...
+                Dim session = StartSession(workspace)
+
+                ' Type a bit in the file
+                Dim caretPosition = workspace.Documents.Single(Function(d) d.CursorPosition.HasValue).CursorPosition.Value
+                Dim textBuffer = workspace.Documents.Single().GetTextBuffer()
+
+                textBuffer.Insert(caretPosition, "Bar")
+                textBuffer.Delete(New Span(caretPosition, "Bar".Length))
+
+                Dim comitted = session.GetTestAccessor().CommitWorker(previewChanges:=False)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/Rename/InlineRenameTests.vb]:
```suggestion
                Dim committed = session.GetTestAccessor().CommitWorker(previewChanges:=False)
```

File: src/EditorFeatures/Test2/Rename/InlineRenameTests.vb [hunk 6]
```diff
@@ -1868,5 +1868,39 @@ class [|$$Test1|]
...
+                Assert.False(comitted)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Rename/InlineRenameTests.vb]:
```suggestion
                Dim committed = session.GetTestAccessor().CommitWorker(previewChanges:=False)
                Assert.False(committed)
```

**Pull request merged**